archive/frames/utils.py

```diff
@@ -156,3 +157,55 @@ def build_nginx_zip_text(frames, directory, uncompress=False):
 
     # Concatenate all lines together into a single string
     return ''.join(ret)
+
+
+def aggregate_raw_sql(frames, timeout=0):
+    frames = frames.values_list(
+        "proposal_id",
+        "site_id",
+        "telescope_id",
+        "instrument_id",
+        "configuration_type",
+        "primary_optical_element",
+    ).distinct().order_by()
+
+    with connections["replica"].cursor() as cursor:
+        distinct_sql, params = frames.query.sql_with_params()
+        params = (timeout,) + params
+        query_sql = f"""
+          SET LOCAL statement_timeout TO %s;
+          SELECT
+            array_agg(DISTINCT site_id) FILTER (WHERE site_id <> ''),
+            array_agg(DISTINCT telescope_id) FILTER (WHERE telescope_id <> ''),
+            array_agg(DISTINCT primary_optical_element) FILTER (WHERE primary_optical_element <> ''),
+            array_agg(DISTINCT instrument_id) FILTER (WHERE instrument_id <> ''),
+            array_agg(DISTINCT configuration_type) FILTER (WHERE configuration_type <> ''),
+            array_agg(DISTINCT proposal_id) FILTER (WHERE proposal_id <> ''),
+            to_char(now() AT TIME ZONE 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS.MSZ')
+          FROM ({distinct_sql}) as t;
```

Comment:
This is some next-level sql-fu! If there are any custom indices you had to make on the table for this to work, can we document them in the README... or better yet, would it be possible to make a django migration that creates them with custom sql if they don't already exist, so new users will have them generated as well?

Reply:
The index that works the best is
```
"frames_frame_agg_desc2" btree (observation_date DESC NULLS LAST, public_date DESC NULLS LAST, proposal_id, site_id, telescope_id, instrument_id, configuration_type, primary_optical_element)
```
I think it could be done with https://docs.djangoproject.com/en/3.2/ref/models/indexes/#expressions